velox/exec/StreamingAggregation.cpp

```diff
@@ -104,11 +115,15 @@ bool equalKeys(
 } // namespace
 
 char* StreamingAggregation::startNewGroup(vector_size_t index) {
-  if (numGroups_ < groups_.size()) {
-    auto group = groups_[numGroups_++];
-    rows_->initializeRow(group, true);
-    storeKeys(group, index);
-    return group;
+  // TODO: Fix the issue that RowContainer::initializeRow would failed when
```

Comment:
Is there a GitHub issue that describes this problem? Should we fix it first before introducing this PR?

Reply:
@mbasmanova Hi Masha, thanks for your review and guidance.
>Is there a GitHub issue that describes this problem?

No, this only happens when enabling disaggregation and the **reuse flag** is true in this PR that the `accumulator.destroy(rows)` in `RowContainer::freeAggregates` would throw an error,
```C++
E20240103 08:39:44.107774 3532646 Exceptions.h:68] Line: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp:426, Function:free, Expression: !header->isFree() , Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Retriable: False
Expression: !header->isFree()
Function: free
File: /Users/macduan/work/native/velox-8/velox/common/memory/HashStringAllocator.cpp
```

>Should we fix it first before introducing this PR?

OK, I will keep investigating this issue first.

Reply:
@mbasmanova Hi Masha, I found the root cause of this failure. I've updated the PR to fix it. Could you please help take a look? Thanks.

- The `DistinctAggregations::extractValues` calls the `aggregate.function->destroy(groups)` to destroy the groups at the end, which would call **`ArrayAggregate::desctroy`**.
- The `StreamingAggregation::startNewGroup` calls `RowContainer::initializeRow`, which calls `RowContainer::freeAggregates`. `RowContainer::freeAggregates` loops the `RowContainer::accumulators_`, and calls `accumulator.destroy(rows)` , which would call **`ArrayAggregate::desctroy`** too.
- The same memory would be freed twice, and cause the error.

```C++
// DistinctAggregations::extractValues
void extractValues(folly::Range<char**> groups, const RowVectorPtr& result)
      override {
    ...
      // Release memory back to HashStringAllocator to allow next
      // aggregate to re-use it.
      aggregate.function->destroy(groups);
    }
  }

char* StreamingAggregation::startNewGroup(vector_size_t index) {
  if (numGroups_ < groups_.size()) {
    auto group = groups_[numGroups_++];
    rows_->initializeRow(group, true);

// RowContainer::initializeRow
char* RowContainer::initializeRow(char* row, bool reuse) {
  if (reuse) {
    auto rows = folly::Range<char**>(&row, 1);
    freeVariableWidthFields(rows);
    freeAggregates(rows);

void RowContainer::freeAggregates(folly::Range<char**> rows) {
  for (auto& accumulator : accumulators_) {
    accumulator.destroy(rows);
  }
}

// ArrayAggregate::desctroy
void destroy(folly::Range<char**> groups) override {
    for (auto group : groups) {
      value<ArrayAccumulator>(group)->elements.free(allocator_);
    }
  }

```

Reply:
Here is how we deal with the same issue in SortedAggregations. Perhaps, same "fix" can be applied to DistinctAggregations::extractValues.

```
void SortedAggregations::extractValues(
    folly::Range<char**> groups,
    const RowVectorPtr& result) {
  raw_vector<int32_t> temp;
...
    for (const auto& aggregate : aggregates) {
      aggregate->function->extractValues(
          groups.data(), groups.size(), &result->childAt(aggregate->output));

      // Release memory back to HashStringAllocator to allow next aggregate to
      // re-use it.
      aggregate->function->destroy(groups);
      // Overwrite empty groups over the destructed groups to keep the container
      // in a well formed state.
      aggregate->function->initializeNewGroups(
          groups.data(),
          folly::Range<const int32_t*>(
              iota(groups.size(), temp), groups.size()));
    }
```

Reply:
It works, thanks for your help.